include/RAJA/index/ListSegment.hpp

```diff
@@ -171,28 +173,35 @@ class TypedListSegment
   TypedListSegment() = delete;
 
   //! Copy constructor for list segment
-  TypedListSegment(const TypedListSegment& other)
-    : m_resource(other.m_resource),
-      m_owned(Unowned), m_data(nullptr), m_size(0)
+  //  As this may be called from a lambda in a
+  //  RAJA method we perform a shallow copy
+  RAJA_HOST_DEVICE TypedListSegment(const TypedListSegment& other)
+    : m_resource(nullptr),
+      m_owned(Unowned), m_data(other.m_data), m_size(other.m_size)
   {
-    bool from_copy_ctor = true;
-    initIndexData(other.m_data, other.m_size, other.m_owned, from_copy_ctor);
+    m_data = const_cast<value_type*>(other.m_data);
+
+    //bool from_copy_ctor = true; //won't be needed
+    //initIndexData(other.m_data, other.m_size, Unowned, from_copy_ctor);
   }
 
   //! Move constructor for list segment
-  TypedListSegment(TypedListSegment&& rhs)
-    : m_resource(rhs.m_resource),
+  RAJA_HOST_DEVICE TypedListSegment(TypedListSegment&& rhs)
+    : 
       m_owned(rhs.m_owned), m_data(rhs.m_data), m_size(rhs.m_size)
   {
     // make the rhs non-owning so it's destructor won't have any side effects
+    m_resource = *rhs.m_resource;
     rhs.m_owned = Unowned;
+    //delete resource;
   }
 
   //! List segment destructor
-  ~TypedListSegment()
+  RAJA_HOST_DEVICE ~TypedListSegment()
   {
     if (m_data != nullptr && m_owned == Owned) {
-      m_resource.deallocate(m_data);
+      m_resource->deallocate(m_data);
+      delete m_resource;
     }
```

Comment:
Move into a new method like `clear()` so it can be reused in copy assignment and move assignment operators.